Refs 2422: Add pending task metrics

jlsherrill opened the pull request:
## Summary
    Adds new metrics:
    * Average queue time of pending task
    * Length in queue for the oldest pending task
    * Number of pending tasks
    Also:
    * Increases default metric calculations to 30 seconds from 5
    * makes interval for metric calculation configurable
    * adds log level override for metrics so you can set the rest of the app to trace while metrics are set to debugging

## Testing steps

on a new db:

```
make run
```
in another tab:
```
make repos-import
go run cmd/external-repos/main.go nightly-jobs
```
This will kick off a ton of tasks, you can monitor with:
```
 curl localhost:9000/metrics | grep task_stats
```

This will update every ~30 seconds.  

## Checklist

- [ ] Tested with snapshotting feature disabled and pulp server URL not configured if appropriate

jlsherrill commented:
note this is currently built ontop of https://github.com/content-services/content-sources-backend/pull/637

will rebase once merged.

jlsherrill commented:
https://issues.redhat.com/browse/HMS-2422

jlsherrill commented:
/retest

swadeley commented:
/retest

swadeley commented:
/retest

rverdile commented on pkg/dao/metrics.go:
```diff
@@ -104,3 +105,40 @@ func (d metricsDaoImpl) PublicRepositoriesFailedIntrospectionCount(ctx context.C
+	return output
+}
+
+func (d metricsDaoImpl) PendingTasksOldestTask(ctx context.Context) float64 {
+	var task models.TaskInfo
+	res := d.db.WithContext(ctx).
+		Model(&models.TaskInfo{}).
+		Where("status = ?", config.TaskStatusPending).Order("queued_at ASC").First(&task)
```
Would we want to alert on this one? IIRC, when a failed task is requeued, the queued_at doesn't reset, so this might cause some false positives. It might make sense to reset the queued_at time.

rverdile commented on pkg/tasks/queue/pgqueue.go:
```diff
@@ -60,7 +60,7 @@ const (
 	sqlRequeue = `
 		UPDATE tasks
-		SET started_at = NULL, token = NULL, status = 'pending', retries = retries + 1
+		SET started_at = NULL, token = NULL, status = 'pending', retries = retries + 1, queued_at = statement_timestamp()
```
I would do the same thing for the `sqlRequeueFailedTasks` query

jlsherrill commented:
added a test to check for queued_at updates.

Also i realized that the 'number of days' until the cdn cert expires was only calculated at startup, so i changed it to be calculated as part of the go routine that updates the metrics.  At startup it was set to zero, so to keep that from firing an alert, i set it to calculate that one metric at startup (and only that metric, because others require the db to be up, which may not actually have happened yet?)

rverdile commented on pkg/config/config.go:
```diff
@@ -181,6 +181,9 @@ type Metrics struct {
 	// metric traffic.
 	Port int `mapstructure:"port"`
+
+	// How often (in seconds) to run queries to collect some metrics
+	CollectionFrequency int `mapstructure:"collection_frequency"`
```
this should also be in config.yaml.example

jlsherrill replied:
yes!

rverdile reviewed (commented):
tested and the numbers look right! just the one small comment

jlsherrill commented:
added
